shared/src/sources/balancer/swap/stable_math.rs

```diff
@@ -0,0 +1,497 @@
+        }
+    }
+
+    Err(Error::StableInvariantDidntConverge)
+}
+
+fn calc_out_given_in(
+    amplification_parameter: U256,
+    balances: &mut [Bfp],
+    token_index_in: usize,
+    token_index_out: usize,
+    token_amount_in: Bfp,
+) -> Result<Bfp, Error> {
+    // Ensure no index error at token indices provided.
+    if token_index_out >= balances.len() || token_index_in >= balances.len() {
+        return Err(Error::InvalidToken);
+    }
+    let invariant = calculate_invariant(amplification_parameter, balances, true)?;
+    balances[token_index_in] = balances[token_index_in].add(token_amount_in)?;
+
+    let final_balance_out = get_token_balance_given_invariant_and_all_other_balances(
+        amplification_parameter,
+        balances,
+        invariant,
+        token_index_out,
+    )?;
+    // No need to use checked arithmetic since `tokenAmountIn` was actually added to the same balance right before
+    // calling `_getTokenBalanceGivenInvariantAndAllOtherBalances` which doesn't alter the balances array.
+    balances[token_index_in] = balances[token_index_in]
+        .sub(token_amount_in)
+        .expect("will not underflow");
+
+    balances[token_index_out]
+        .sub(final_balance_out)?
+        .sub(Bfp::from_wei(1.into()))
+}
+
+fn calc_in_given_out(
+    amplification_parameter: U256,
+    balances: &mut [Bfp],
+    token_index_in: usize,
+    token_index_out: usize,
+    token_amount_out: Bfp,
+) -> Result<Bfp, Error> {
+    // Ensure no index error at token indices provided.
+    if token_index_out >= balances.len() || token_index_in >= balances.len() {
+        return Err(Error::InvalidToken);
+    }
+    let invariant = calculate_invariant(amplification_parameter, balances, true)?;
+    balances[token_index_out] = balances[token_index_out].sub(token_amount_out)?;
+
+    let final_balance_in = get_token_balance_given_invariant_and_all_other_balances(
+        amplification_parameter,
+        balances,
+        invariant,
+        token_index_in,
+    )?;
+
+    // No need to use checked arithmetic since `tokenAmountOut` was actually subtracted from the same balance right
+    // before calling `_getTokenBalanceGivenInvariantAndAllOtherBalances` which doesn't alter the balances array.
+    balances[token_index_out] = balances[token_index_out]
+        .add(token_amount_out)
+        .expect("Will not overflow");
+
+    final_balance_in
+        .sub(balances[token_index_in])?
+        .add(Bfp::from_wei(1.into()))
+}
+
+fn get_token_balance_given_invariant_and_all_other_balances(
+    amplification_parameter: U256,
+    balances: &[Bfp],
+    invariant: U256,
+    token_index: usize,
+) -> Result<Bfp, Error> {
+    // Rounds result up overall
+    let num_tokens_usize = balances.len();
+    let num_tokens = U256::from(num_tokens_usize);
+    let amp_times_total = amplification_parameter.bmul(num_tokens)?;
+    let mut sum = balances[0].as_uint256();
+    let mut p_d = balances[0].as_uint256().bmul(num_tokens)?;
```

Comment:
Could panic if balances is empty. Again I don't know if this can happen in practice just wanted to note it.

Reply:
I'll add it to the index verification at the top.

Reply:
I don't think we need index verification here as this is an internal function that should only get called by `calc_*_given_*` which would return an error if the balances slice is empty (since there is no `token_*_index` for which `token_*_index < bal.len() && bal.len() == 0`).

I think just adding a note that `panics if balances is empty` is fine.

Reply:
gefixed!